Snowflake.Data/Core/Session/SessionPool.cs

```diff
@@ -95,68 +103,170 @@ internal SFSession GetSession(string connStr, SecureString password)
         {
             s_logger.Debug("SessionPool::GetSession");
             if (!_pooling)
-                return NewSession(connStr, password);
-            SFSession session = GetIdleSession(connStr);
-            return session ?? NewSession(connStr, password);
+                return NewSession(connStr, password, _noPoolingSessionCreationTokenCounter.NewToken());
+            var sessionOrCreateToken = GetIdleSession(connStr);
+            if (sessionOrCreateToken.Session != null)
+            {
+                _sessionPoolEventHandler.OnSessionProvided(this);
+            }
+            return sessionOrCreateToken.Session ?? NewSession(connStr, password, sessionOrCreateToken.SessionCreationToken);
         }
         
-        internal Task<SFSession> GetSessionAsync(string connStr, SecureString password, CancellationToken cancellationToken)
+        internal async Task<SFSession> GetSessionAsync(string connStr, SecureString password, CancellationToken cancellationToken)
```

Comment:
This method was not async, and was providing a Task for the app to wait for results. 
why change the API?

Reply:
this was a bug that SnowflakeDbConnection.OpenAsync in case of exception did not handle faulted task but instead  the exception was thrown without converting to a faulted task. I added test TestWaitForTheIdleConnectionWhenExceedingMaxConnectionsLimitAsync to cover it.